nrf-hal-common/src/uarte.rs

```diff
@@ -408,18 +422,324 @@ pub enum Error {
+    written: usize,
+}
+
+/// Interface for the RX part of a UART instance that can be used independently of the TX part.
+pub struct UarteRx<'a, T>
+where
+    T: Instance,
+{
+    _marker: core::marker::PhantomData<T>,
+    rx_buf: &'a mut [u8],
+}
+
+impl<'a, T> UarteTx<'a, T>
+where
+    T: Instance,
+{
+    fn new(tx_buf: &'a mut [u8]) -> Result<UarteTx<'a, T>, Error> {
+        slice_in_ram_or(tx_buf, Error::BufferNotInRAM)?;
+        if tx_buf.len() == 0 {
+            return Err(Error::TxBufferTooSmall);
+        }
+
+        if tx_buf.len() > EASY_DMA_SIZE {
+            return Err(Error::TxBufferTooLong);
+        }
+
+        Ok(UarteTx {
+            _marker: core::marker::PhantomData,
+            tx_buf,
+            written: 0,
+        })
+    }
+}
+
+impl<'a, T> UarteRx<'a, T>
+where
+    T: Instance,
+{
+    fn new(rx_buf: &'a mut [u8]) -> Result<UarteRx<'a, T>, Error> {
+        slice_in_ram_or(rx_buf, Error::BufferNotInRAM)?;
+        if rx_buf.len() == 0 {
+            return Err(Error::RxBufferTooSmall);
+        }
+
+        if rx_buf.len() > EASY_DMA_SIZE {
+            return Err(Error::RxBufferTooLong);
+        }
+
+        Ok(UarteRx {
+            _marker: core::marker::PhantomData,
+            rx_buf,
+        })
+    }
+}
+
+impl<'a, T> Drop for UarteTx<'a, T>
+where
+    T: Instance,
+{
+    fn drop(&mut self) {
+        let uarte = unsafe { &*T::ptr() };
+
+        let in_progress = uarte.events_txstarted.read().bits() == 1;
+        // Stop any ongoing transmission
+        if in_progress {
+            uarte.tasks_stoptx.write(|w| unsafe { w.bits(1) });
+
+            // Wait for transmitter is stopped.
+            while uarte.events_txstopped.read().bits() == 0 {}
+
+            // Reset events
+            uarte.events_endtx.reset();
+            uarte.events_txstopped.reset();
+
+            // Ensure the above is done
+            compiler_fence(SeqCst);
+        }
+    }
+}
+
+impl<'a, T> Drop for UarteRx<'a, T>
+where
+    T: Instance,
+{
+    fn drop(&mut self) {
+        let uarte = unsafe { &*T::ptr() };
+
+        let in_progress = uarte.events_rxstarted.read().bits() == 1;
+        // Stop any ongoing reception
+        if in_progress {
+            uarte.tasks_stoprx.write(|w| unsafe { w.bits(1) });
+
+            // Wait for receive to be done to ensure memory is untouched.
+            while uarte.events_rxto.read().bits() == 0 {}
+
+            uarte.events_rxto.reset();
+
+            // Flush DMA
+            uarte.tasks_flushrx.write(|w| unsafe { w.bits(1) });
+
+            // Wait for the flush to complete.
+            while uarte.events_endrx.read().bits() == 0 {}
+
+            // Reset events
+            uarte.events_endrx.reset();
+
+            // Ensure the above is done
+            compiler_fence(SeqCst);
+        }
+    }
+}
+
+pub mod serial {
+
+    ///! Implementation of the embedded_hal::serial::* and embedded_hal::blocking::serial::*  traits for UartTx and UartRx.
```

Comment:
```suggestion
/// Implementation of the embedded_hal::serial::* and embedded_hal::blocking::serial::*  traits for UartTx and UartRx.
pub mod serial {
```

Reply:
This diff doesn't seem to have been applied

Reply:
Oh actually the whole module should be removed and the impls put into the parent, there's nothing else in there

Reply:
It was only intended to contain the embedded_hal::serial import. I can move it to the parent module, thats fine.

Reply:
fixed.